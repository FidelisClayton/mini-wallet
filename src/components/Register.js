import React, { Component } from 'react'
import { css } from 'emotion'
import md5 from 'md5'

import Button from './Button'
import LabeledInput from './LabeledInput'
import Card from './Card'

const styles = css({
  transition: `
    500ms ease margin-top 300ms,
    500ms ease transform
  `,

  '&.register--inactive': {
    position: 'absolute',
    transform: 'scale(0.90)',
    marginTop: '45px',
    paddingBottom: 0,
    zIndex: 0,
  }
})

const initialState = {
  data: {
    email: '',
    password: '',
    passwordAgain: ''
  },
  errors: []
}

export default class Register extends Component {
  state = initialState

  validatePasswords = (password, passwordAgain) => {
    if(passwordAgain !== '') {
      if (password === passwordAgain) {
        return
      } else {
        return {
          field: 'passwordAgain',
          message: 'As senhas não são iguais'
        }
      }
    }

    return
  }

  validateFields = (data) => {
    const errors = [
      this.validatePasswords(data.password, data.passwordAgain)
    ].filter(error => !!error)

    return errors
  }

  handleSubmit = (event) => {
    event.preventDefault()

    const data = this.state.data

    this.props.onSubmit({
      email: data.email,
      password: md5(data.password),
      onboardingComplete: false
    })
  }

  handleInputChange = (inputName) => (event) => {
    const value = event.target.value

    const updatedData = {
      ...this.state.data,
      [inputName]: value
    }

    const errors = this.validateFields(updatedData)

    this.setState({
      data: {
        ...this.state.data,
        [inputName]: event.target.value
      },
      errors
    })
  }

  getErrorForField = (fieldName) => {
    const error = this.state.errors.find(error => error.field === fieldName) || {}

    return error.message
  }

  reset = () => {
    this.setState(initialState)
  }

  get classNames () {
    const classNames = [
      'register',
      !this.props.active && 'register--inactive',
      styles
    ]
      .filter(className => !!className)
      .join(' ')

    return classNames
  }

  render () {
    return (
      <Card className={this.classNames}>
        <h4 className="card__title">
          Criar Conta
        </h4>

        <LabeledInput
          label="Email"
          inputType="email"
          onChange={this.handleInputChange('email')}
          error={this.getErrorForField('email')}
          inputProps={{
            defaultValue: this.state.data.email
          }}
        />

        <LabeledInput
          label="Senha"
          inputType="password"
          onChange={this.handleInputChange('password')}
          error={this.getErrorForField('password')}
          inputProps={{
            defaultValue: this.state.data.password
          }}
        />

        <LabeledInput
          label="Repita a senha"
          inputType="password"
          onChange={this.handleInputChange('passwordAgain')}
          error={this.getErrorForField('passwordAgain')}
          inputProps={{
            defaultValue: this.state.data.passwordAgain
          }}
        />

        { this.props.active && (
          <div className="card__submit">
            <Button
              type="button"
              onClick={this.handleSubmit}
            >
              Registrar
            </Button>
          </div>
        )}

        { !this.props.active && (
          <h4
            className="card__title card__title--bottom"
            onClick={this.props.onClick}
          >
            Criar Conta
          </h4>
        )}
      </Card>
    )
  }
}
