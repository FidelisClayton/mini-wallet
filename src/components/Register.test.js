import React from 'react'
import { shallow } from 'enzyme'

import Register from './Register'

describe('Component: <Register />', () => {
  it('should properly render', () => {
    const wrapper = shallow(
      <Register
        onClick={jest.fn()}
        onSubmit={jest.fn()}
        active={true}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should handle title click', () => {
    const onClick = jest.fn()

    const wrapper = shallow(
      <Register
        onClick={onClick}
        onSubmit={jest.fn()}
        active={false}
      />
    )

    wrapper.find('.card__title--bottom').simulate('click')
    expect(onClick.mock.calls.length).toEqual(1)
  })
})
