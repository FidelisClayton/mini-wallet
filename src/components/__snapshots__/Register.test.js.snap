// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: <Register /> should properly render 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Register
    active={true}
    onClick={[Function]}
    onSubmit={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <h4
          className="card__title"
>
          Criar Conta
</h4>,
        <LabeledInput
          error={undefined}
          inputProps={
                    Object {
                              "defaultValue": "",
                            }
          }
          inputType="email"
          label="Email"
          onChange={[Function]}
/>,
        <LabeledInput
          error={undefined}
          inputProps={
                    Object {
                              "defaultValue": "",
                            }
          }
          inputType="password"
          label="Senha"
          onChange={[Function]}
/>,
        <LabeledInput
          error={undefined}
          inputProps={
                    Object {
                              "defaultValue": "",
                            }
          }
          inputType="password"
          label="Repita a senha"
          onChange={[Function]}
/>,
        <div
          className="card__submit"
>
          <Button
                    onClick={[Function]}
                    type="button"
          >
                    Registrar
          </Button>
</div>,
        false,
      ],
      "className": "register css-wa52u1",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Criar Conta",
          "className": "card__title",
        },
        "ref": null,
        "rendered": "Criar Conta",
        "type": "h4",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "error": undefined,
          "inputProps": Object {
            "defaultValue": "",
          },
          "inputType": "email",
          "label": "Email",
          "onChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "error": undefined,
          "inputProps": Object {
            "defaultValue": "",
          },
          "inputType": "password",
          "label": "Senha",
          "onChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "error": undefined,
          "inputProps": Object {
            "defaultValue": "",
          },
          "inputType": "password",
          "label": "Repita a senha",
          "onChange": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <Button
            onClick={[Function]}
            type="button"
>
            Registrar
</Button>,
          "className": "card__submit",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": "Registrar",
            "onClick": [Function],
            "type": "button",
          },
          "ref": null,
          "rendered": "Registrar",
          "type": [Function],
        },
        "type": "div",
      },
      false,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <h4
            className="card__title"
>
            Criar Conta
</h4>,
          <LabeledInput
            error={undefined}
            inputProps={
                        Object {
                                    "defaultValue": "",
                                  }
            }
            inputType="email"
            label="Email"
            onChange={[Function]}
/>,
          <LabeledInput
            error={undefined}
            inputProps={
                        Object {
                                    "defaultValue": "",
                                  }
            }
            inputType="password"
            label="Senha"
            onChange={[Function]}
/>,
          <LabeledInput
            error={undefined}
            inputProps={
                        Object {
                                    "defaultValue": "",
                                  }
            }
            inputType="password"
            label="Repita a senha"
            onChange={[Function]}
/>,
          <div
            className="card__submit"
>
            <Button
                        onClick={[Function]}
                        type="button"
            >
                        Registrar
            </Button>
</div>,
          false,
        ],
        "className": "register css-wa52u1",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Criar Conta",
            "className": "card__title",
          },
          "ref": null,
          "rendered": "Criar Conta",
          "type": "h4",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "error": undefined,
            "inputProps": Object {
              "defaultValue": "",
            },
            "inputType": "email",
            "label": "Email",
            "onChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "error": undefined,
            "inputProps": Object {
              "defaultValue": "",
            },
            "inputType": "password",
            "label": "Senha",
            "onChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "error": undefined,
            "inputProps": Object {
              "defaultValue": "",
            },
            "inputType": "password",
            "label": "Repita a senha",
            "onChange": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <Button
              onClick={[Function]}
              type="button"
>
              Registrar
</Button>,
            "className": "card__submit",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": "Registrar",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": "Registrar",
            "type": [Function],
          },
          "type": "div",
        },
        false,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
